document/AnnexE-Determinism

```diff
@@ -0,0 +1,52 @@
+# Annex E (informative) Producing deterministic SARIF log files { - #Annex-E }
+
+In certain circumstances, it is desirable for an analysis tool to produce deterministic output;
+that is, for it to produce identical output when run repeatedly over identical inputs.
+
+Certain build systems provide an example of when this is desirable.
+Consider a build system that caches the results of each build step.
+If the build is rerun, and the inputs to the step are identical
+(which the build system might determine, for example, by comparing timestamps,
+or by computing a hash of the inputs to the step and storing it along with the output from the step),
+then the build system can save time by not re-running the step, and simply
+using the existing outputs.
+
+In the case of SARIF, one could imagine a sequence of build steps where Steps A
+B, and C each run an analysis tool on a different set of targets, producing log files
+A.sarif, B.sarif, and C.sarif, and then build Step D performs an analysis on the aggregate
+of those log files. If the targets analyzed in Step B change
+but the targets analyzed in steps A and C do not, and if the contents of the
+SARIF log file are deterministic, then when the build is re-run, only Steps B and D
+need be performed.
+
+If it is desired to produce deterministic output, the following elements of the SARIF format
+should not be used. All of these elements are optional in the format:
+
+* `invocation.startTime`
+* `invocation.endTime`
+* `invocation.processId`
+* `invocation.machine`
+* `invocation.account`
+* `invocation.fileName` (because `fileName` is specified as being an absolute path, and
+tools might be stored in different directories on different machines)
+* `invocation.workingDirectory`
+* `invocation.environmentVariables`
+* The use of absolute file paths in `invocation.commandLine`
+(because builds performed on different machines might use a different root directory)
+* `annotatedCodeLocation.threadId`
+* `notification.threadId`
+* `notification.time`
+* `result.baselineState` (because it depends on a previous run.
+A result whose `baselineState` was `"new"` in a previous run would be `"existing"`
+in the next run.)
+* `run.id`
+* `run.automationId`
+* `run.baselineId`
```

Comment:
(I know you said "could be" stable; I'll address that at the end.)

A common example of `run.automationId` (this is the example given in the spec) would be the build id:

```
"automationId": "Build-14.0.1.2-20160518-15:48:02",
```

So if the same tool is run on the same inputs in two successive daily builds, the `automationId`s would differ.

As for `run.id`, the spec defines it to be

> a string which uniquely identifies the run

So it _has_ to change from run to run.

Now since `run.baselineId` is defined to be

> a string which shall match the `id` property of some previous run

So if you run the tool twice against the same baseline, then `baselineId` will be the same. But in the common scenario where your baseline is yesterday's build, then `baselineId` will be different each time.

In short: I don't think either `run.automationId` or `run.id` could be stable. `run.baselineId` indeed could be stable in a particular scenario. But all in all, I think the guidance I currently give, which is that these elements "should not be used" if you want deterministic output, is good.